pkg/storage/stores/tsdb/store.go

```diff
@@ -10,16 +13,39 @@ import (
 	"github.com/grafana/loki/pkg/storage/stores/indexshipper"
 	"github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads"
 	"github.com/grafana/loki/pkg/storage/stores/series"
+	"github.com/grafana/loki/pkg/storage/stores/tsdb/index"
 	util_log "github.com/grafana/loki/pkg/util/log"
 )
 
+// we do not need to build store for each schema config since we do not do any schema specific handling yet.
+// If we do need to do schema specific handling, it would be a good idea to abstract away the handling since
+// running multiple head managers would be complicated and wasteful.
+var storeInstance *store
+
+type store struct {
+	indexWriter IndexWriter
+	indexStore  series.IndexStore
+}
+
+// NewStore creates a new store if not initialized already.
+// Each call to NewStore will always build a new stores.ChunkWriter even if the store was already initialized since
+// fetcher.Fetcher instances could be different due to periodic configs having different types of object storage configured
+// for storing chunks.
 func NewStore(indexShipperCfg indexshipper.Config, p config.PeriodConfig, f *fetcher.Fetcher,
 	objectClient client.ObjectClient, limits downloads.Limits, reg prometheus.Registerer) (stores.ChunkWriter, series.IndexStore, error) {
-	var (
-		nodeName = indexShipperCfg.IngesterName
-		dir      = indexShipperCfg.ActiveIndexDirectory
-	)
-	tsdbMetrics := NewMetrics(reg)
+	if storeInstance == nil {
```

Comment:
I understand but really dislike the side effect of mutating the global variable within the `New` function. Can we handle this in a cleaner fashion by passing the `IndexWriter` as an argument to this function?